src/Microsoft.AspNet.Razor.Runtime/BufferedTextWriter.cs

```diff
@@ -0,0 +1,189 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNet.Razor.Runtime
+{
+    /// <summary>
+    /// TextWriter which is used to write <see cref="Runtime.BufferedStringCollection"/>.
+    /// </summary>
+    public class BufferedTextWriter : TextWriter
+    {
+        private static readonly Task _completedTask = Task.FromResult(0);
+        private readonly Encoding _encoding;
+
+        /// <summary>
+        /// Creates a new instance of <see cref="BufferedTextWriter"/>.
+        /// </summary>
+        /// <param name="encoding">The character <see cref="Encoding"/> in which the output is written.</param>
+        public BufferedTextWriter(Encoding encoding)
+        {
+            _encoding = encoding;
+            BufferedStringCollection = new BufferedStringCollection();
+        }
+        
+        public BufferedStringCollection BufferedStringCollection { get; set; }
+
+        /// <inheritdoc />
+        public override Encoding Encoding
+        {
+            get { return _encoding; }
+        }
+
+        /// <inheritdoc />
+        public override void Write(char value)
+        {
+            BufferedStringCollection.Append(value.ToString());
+        }
+
+        /// <inheritdoc />
+        public override void Write([NotNull] char[] buffer, int index, int count)
+        {
+            if (index < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(index));
+            }
+            if (count < 0 || (buffer.Length - index < count))
+            {
+                throw new ArgumentOutOfRangeException(nameof(count));
+            }
+
+            BufferedStringCollection.Buffer.Add(buffer, index, count);
```

Comment:
This is concerning and ties into my comment above about it needing to be in `RazorRuntime`. Is there any reason this `TextWriter` can't just utilize the `BufferEntryCollection` alone?